deps/src/caller.cpp

```diff
@@ -0,0 +1,340 @@
+        jl_arrayset(current_return, jl_box_int64(types[i].first), 0);
+        jl_arrayset(current_return,
+                    reinterpret_cast<jl_value_t *>(
+                        jl_symbol(types[i].second.c_str())),
+                    1);
+        jl_arrayset(return_array,
+                    reinterpret_cast<jl_value_t *>(current_return), i);
+    }
+    return reinterpret_cast<jl_value_t *>(return_array);
+}
+
+static void initialize_jl_c_types(jl_value_t* module_value)
+{
+    jl_module_t* module = reinterpret_cast<jl_module_t*>(module_value);
+    jl_int64_vector_type =
+        jl_apply_array_type((jl_value_t *)jl_int64_type, 1);
+    jl_int64_matrix_type =
+        jl_apply_array_type((jl_value_t *)jl_int64_type, 2);
+    jl_singular_number_type = jl_get_global(module,jl_symbol("number"));
+    jl_singular_poly_type = jl_get_global(module,jl_symbol("poly"));
+    jl_singular_ring_type = jl_get_global(module,jl_symbol("ring"));
+    jl_singular_ideal_type = jl_get_global(module,jl_symbol("ideal"));
+    jl_singular_matrix_type = jl_get_global(module,jl_symbol("ip_smatrix"));
+    jl_singular_bigint_type = jl_get_global(module,jl_symbol("__mpz_struct"));
+    jl_singular_bigintmat_type = jl_get_global(module,jl_symbol("bigintmat"));
+    jl_singular_bigintmat_type = jl_get_global(module,jl_symbol("sip_smap"));
+    jl_singular_resolution_type = jl_get_global(module,jl_symbol("resolvente"));
+}
+
+static inline void * get_ptr_from_cxxwrap_obj(jl_value_t * obj)
+{
+    return *reinterpret_cast<void **>(obj);
+}
+
+// Safe way
+// void* get_ptr_from_cxxwrap_obj(jl_value_t* obj){
+//     return jl_unbox_voidpointer(jl_get_field(obj,"cpp_object"));
+// }
+
+jl_value_t * intvec_to_jl_array(intvec * v)
+{
+    int          size = v->length();
+    jl_array_t * result = jl_alloc_array_1d(jl_int64_vector_type, size);
+    int *        v_content = v->ivGetVec();
+    for (int i = 0; i < size; i++) {
+        jl_arrayset(result, jl_box_int64(static_cast<int64_t>(v_content[i])),
+                    i);
+    }
+    return reinterpret_cast<jl_value_t *>(result);
+}
+
+jl_value_t * intmat_to_jl_array(intvec * v)
+{
+    int rows = v->rows();
+    int cols = v->cols();
+    jl_array_t * result = jl_alloc_array_2d(jl_int64_matrix_type, rows, cols);
+    int64_t* result_ptr = reinterpret_cast<int64_t*>jl_array_data(result);
+    for (int i = 0; i < rows; i++) {
+        for( int j = 0; j < cols; j++){
+            result_ptr[ j + (i*cols) ] = IMATELEM(*v,i,j);
+        }
+    }
+    return reinterpret_cast<jl_value_t *>(result);
+}
+
+intvec * jl_array_to_intvec(jl_value_t * array_val)
+{
+    jl_array_t * array = reinterpret_cast<jl_array_t *>(array_val);
+    int          size = jl_array_len(array);
+    intvec *     result = new intvec(size);
+    int *        result_content = result->ivGetVec();
+    for (int i = 0; i < size; i++) {
+        result_content[i] =
+            static_cast<int>(jl_unbox_int64(jl_arrayref(array, i)));
```

Comment:
Well, making it only work with `Int32` is IMHO out of question, as this will be super inconvenient. I could provide both versions, as @fingolfin suggests, and raise an error if there is an overflow. I can either check on the Julia or the C++-side, it does not matter.